trade-finance-manager-ui/server/controllers/utilisation-reports/record-corrections/initiate-record-correction-request/index.ts

```diff
@@ -0,0 +1,108 @@
+import { getPremiumPaymentsFacilityIdQueryFromReferer } from '../../../../helpers/get-premium-payments-facility-id-query-from-referer';
+import { InitiateRecordCorrectionRequestErrorKey } from '../../../../types/premium-payments-tab-error-keys';
+import { INITIATE_RECORD_CORRECTION_ERROR_KEY } from '../../../../constants/premium-payment-tab-error-keys';
+import { mapCheckedCheckboxesToRecord } from '../../../../helpers/map-checked-checkboxes-to-record';
+
+export type PostInitiateRecordCorrectionRequest = CustomExpressRequest<{
+  reqBody: PremiumPaymentsTableCheckboxSelectionsRequestBody;
+}>;
+
+/**
+ * Redirects the user to the premium payments page with an error message.
+ * @param req - The Express request object.
+ * @param res - The Express response object.
+ * @param reportId - The ID of the utilisation report.
+ * @param initiateRecordCorrectionRequestError - The error key to be set in the session.
+ * @param checkedCheckboxIds - An object representing the checked checkbox IDs.
+ * @returns The redirect response.
+ */
+const redirectToPremiumPaymentsWithError = (
+  req: PostInitiateRecordCorrectionRequest,
+  res: Response,
+  reportId: string,
+  initiateRecordCorrectionRequestError: InitiateRecordCorrectionRequestErrorKey,
+  checkedCheckboxIds: Record<string, true | undefined> = {},
+) => {
+  req.session.initiateRecordCorrectionRequestErrorKey = initiateRecordCorrectionRequestError;
+  req.session.checkedCheckboxIds = checkedCheckboxIds;
+
+  const premiumPaymentsFacilityId = getPremiumPaymentsFacilityIdQueryFromReferer(req.headers.referer);
+  return res.redirect(axios.getUri({ url: `/utilisation-reports/${reportId}`, params: { premiumPaymentsFacilityId } }));
+};
+
+/**
+ * Controller for the POST initiate record correction request route.
+ *
+ * Validates selected fee records for creating a record correction request.
+ *
+ * Redirects to the premium payments tab with the appropriate error message
+ * if the selections are not valid for record correction request creation.
+ *
+ * Redirects to the create record correction request page for the selected
+ * fee record if the selections are valid for record correction request creation.
+ *
+ * @param req - The request object
+ * @param res - The response object
+ */
+export const postInitiateRecordCorrectionRequest = (req: PostInitiateRecordCorrectionRequest, res: Response) => {
+  try {
+    const { reportId } = req.params;
+
+    const checkedCheckboxIds = getPremiumPaymentsCheckboxIdsFromObjectKeys(req.body);
+
+    if (checkedCheckboxIds.length === 0) {
+      return redirectToPremiumPaymentsWithError(
+        req,
+        res,
+        reportId,
+        INITIATE_RECORD_CORRECTION_ERROR_KEY.NO_FEE_RECORDS_SELECTED,
+        mapCheckedCheckboxesToRecord(checkedCheckboxIds),
+      );
+    }
+
+    if (checkedCheckboxIds.length > 1) {
+      return redirectToPremiumPaymentsWithError(
+        req,
+        res,
```

Comment:
Can we pass specific properties to `redirectToPremiumPaymentsWithError` please, or maybe the `req.session` object, instead of passing the whole req and res objects around

Reply:
Can I ask why? I think it is quite nice like this and I think it helps keep it clear

Reply:
Passing req/res around is not a good practice / code smell. In this instance it's also overkill as we're passing 2x huge req/res objects to another function when that function only needs 2x specific things

Also means you have to hunt around in another function, to see what it needs/uses from req/res, rather than passing in specific params

Reply:
Ok I tried it but I get multiple ts lint errors, one for 'no-param-reassign' and one for 'unbound-method'.

<img width="1080" alt="Screenshot 2024-11-18 at 17 22 05" src="https://github.com/user-attachments/assets/0c2f7567-9b38-4071-9a79-59099dbc251c">

<img width="1513" alt="Screenshot 2024-11-18 at 17 21 55" src="https://github.com/user-attachments/assets/b0fd58a1-563b-4643-a835-969ed0c6db35">

I could remove this function and just inline these lines, would be slight repetition but would remove the passing of the req, res objects to another function?